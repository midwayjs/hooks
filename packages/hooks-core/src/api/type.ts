import type { Merge } from 'type-fest'
import { AsyncFunction } from '../types/common'

export type ArrayToObject<T, R = {}> = T extends [infer First, ...infer Rest]
  ? First extends PromiseLike<infer PromiseValue>
    ? PromiseValue
    : First extends object
    ? Merge<First, ArrayToObject<Rest, R>>
    : ArrayToObject<Rest, R>
  : R

export type ApiRunner<
  Input extends object | void,
  Handler extends AsyncFunction
> = (
  ...args: Input extends void
    ? Parameters<Handler>
    : [...args: Parameters<Handler>, input: Input]
) => ReturnType<Handler>

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export type MetadataHelper = {
  setMetadata: <T = any>(key: any, value: T) => void
  getMetadata: <T = any>(key: any) => T
}

export type ExecuteHelper = {
  result?: any
  getInputArguments?: () => any[]
}

export type Operator<Input> = {
  name: string
  type?: Input
  input?: boolean
  metadata?: (helper: MetadataHelper) => void
  execute?: (helper: ExecuteHelper, next: () => Promise<any>) => Promise<void>
}

export type ExtractInputType<T> = {
  [key in keyof T]: T[key] extends Operator<any> ? T[key]['type'] : void
}

export type BaseTrigger = {
  type: string
  requestClient?: {
    fetcher: string
    client: string
  }
  [key: string]: any
}
