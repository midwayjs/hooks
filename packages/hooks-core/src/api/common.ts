import { isFunction } from '../common'

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export const isApiFunction = (target: any): boolean => {
  return (
    isFunction(target) && !!Reflect.getMetadata(OperatorType.Trigger, target)
  )
}
