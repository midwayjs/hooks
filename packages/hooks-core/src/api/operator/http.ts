import { BaseTrigger, MetadataHelper, Operator, OperatorType } from '../type'

export const HttpTriggerType = 'HTTP'

export enum HttpMetadata {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAMS = 'Http_Params',
  HEADERS = 'Http_Headers',
  RESPONSE = 'Http_Response',
}

export enum ResponseMetaType {
  CODE = 'Http_Response_Code',
  HEADER = 'Http_Response_Header',
  CONTENT_TYPE = 'Http_Response_ContentType',
  REDIRECT = 'Http_Response_Redirect',
}

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
  HEAD = 'HEAD',
  OPTIONS = 'OPTIONS',
  ALL = 'ALL',
}

export interface HttpTrigger extends BaseTrigger {
  type: typeof HttpTriggerType
  method: HttpMethod
  path?: string
}

export type HttpInputMetadata = {
  query?: Record<string, string>
  headers?: Record<string, string>
  params?: Record<string, string>
}

function createHttpMethodOperator(method: HttpMethod) {
  return (path?: string) => {
    return {
      name: method,
      metadata({ setMetadata }) {
        setMetadata<HttpTrigger>(OperatorType.Trigger, {
          type: HttpTriggerType,
          method,
          path,
          requestClient: {
            fetcher: 'http',
            client: '@midwayjs/rpc',
          },
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHttpMethodOperator(HttpMethod.ALL)
export const Get = createHttpMethodOperator(HttpMethod.GET)
export const Post = createHttpMethodOperator(HttpMethod.POST)
export const Delete = createHttpMethodOperator(HttpMethod.DELETE)
export const Put = createHttpMethodOperator(HttpMethod.PUT)
export const Patch = createHttpMethodOperator(HttpMethod.PATCH)
export const Head = createHttpMethodOperator(HttpMethod.HEAD)
export const Options = createHttpMethodOperator(HttpMethod.OPTIONS)

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpMetadata.QUERY,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.QUERY, true)
    },
  }
}

export function Params<T extends Record<string, string>>(): Operator<{
  params: T
}> {
  return {
    name: HttpMetadata.PARAMS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.PARAMS, true)
    },
  }
}

export function Headers<T extends Record<string, string>>(): Operator<{
  headers: T
}> {
  return {
    name: HttpMetadata.HEADERS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.HEADERS, true)
    },
  }
}

export type ResponseMetaData = {
  type: ResponseMetaType
  code?: number
  header?: {
    key: string
    value: string
  }
  url?: string
  contentType?: string
}

export function HttpCode(code: number): Operator<void> {
  return {
    name: 'HttpCode',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CODE, { code })
    },
  }
}

export function SetHeader(key: string, value: string): Operator<void> {
  return {
    name: 'SetHeader',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.HEADER, {
        header: {
          key,
          value,
        },
      })
    },
  }
}

export function Redirect(url: string, code?: number): Operator<void> {
  return {
    name: 'Redirect',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.REDIRECT, {
        url,
        code,
      })
    },
  }
}

export function ContentType(contentType: string): Operator<void> {
  return {
    name: 'ContentType',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CONTENT_TYPE, {
        contentType,
      })
    },
  }
}

function setResponseMetaData(
  helper: MetadataHelper,
  type: ResponseMetaType,
  value: Partial<ResponseMetaData>
) {
  const responseMetaData =
    helper.getMetadata<ResponseMetaData[]>(HttpMetadata.RESPONSE) || []

  helper.setMetadata(HttpMetadata.RESPONSE, [
    ...responseMetaData,
    {
      type,
      ...value,
    },
  ])
}
