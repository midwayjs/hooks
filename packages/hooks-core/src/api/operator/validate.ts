import { z } from 'zod'
import { validateFunction } from '../../'
import { Operator } from '../type'

/**
 * Validate input arguments.
 * Default lib is zod.
 */
export type Validator = (schemas: any[], input: any[]) => any | Promise<any>

let validator: Validator = (schemas: any, input: any[]) => {
  const tuple = z.tuple(schemas as any)
  return tuple.parseAsync(input)
}

export function setValidator(newValidator: Validator) {
  validateFunction(newValidator, 'newValidator')
  validator = newValidator
}

export function Validate(...schemas: any[]): Operator<void> {
  return {
    name: 'Validate',
    async execute({ getInputArguments }, next) {
      validateFunction(validator, 'validator')
      const inputs = getInputArguments()
      await validator(schemas, inputs)
      return next()
    },
  }
}
