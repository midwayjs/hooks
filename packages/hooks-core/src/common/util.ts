import parseArgs from 'fn-args'
import debug from 'debug'
import isClass from 'is-class'

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return typeof fn === 'function' && parseArgs(fn).length === 1 && !isClass(fn)
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}

export function createDebug(namespace: string) {
  return debug(namespace)
}
