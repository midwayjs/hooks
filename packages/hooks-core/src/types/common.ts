import { BaseTrigger, HttpMethod } from '..'

export type FunctionId = string

export type AsyncFunction = (...args: any[]) => Promise<any>

export interface ApiFunction extends AsyncFunction {}

export type ApiConfig = {
  middleware?: HooksMiddleware[] | any[]
}

export type ApiModule = {
  config?: ApiConfig
  [index: string]: ApiFunction | any
}

export type HooksMiddleware = (next: () => any | Promise<any>) => any

export type RawRequestOptions = {
  trigger: BaseTrigger
  args?: any
  metadata?: any
}

export type FileRecord = Record<string, File | FileList>

export type HttpRequestOptions = {
  url: string
  method: HttpMethod
  data?: { args: any[] } | FormData

  // query & headers
  query?: Record<string, string>
  headers?: Record<string, string>
  files?: FileRecord
}

export type RequestRoute<T = any> = {
  trigger: BaseTrigger & T
  functionId: FunctionId
  useInputMetadata: boolean
}

export type RequestArgs<
  Trigger,
  InputMetaData = void
> = InputMetaData extends void
  ? [...args: any[], route: RequestRoute<Trigger>]
  : [...args: any[], inputMetadata: InputMetaData, route: RequestRoute<Trigger>]
