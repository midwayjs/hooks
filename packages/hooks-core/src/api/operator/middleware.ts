import { HooksMiddleware } from '../../index'
import { Operator } from '../type'
import { OperatorType } from '../common'

export function Middleware(
  middleware: HooksMiddleware[] | any[]
): Operator<void>
export function Middleware(...args: HooksMiddleware[] | any[]): Operator<void>
export function Middleware(...middlewares): Operator<void> {
  let middleware: HooksMiddleware[] = middlewares
  if (Array.isArray(middlewares[0])) {
    middleware = middlewares[0]
  }

  return {
    name: 'Middleware',
    metadata({ setMetadata }) {
      setMetadata(OperatorType.Middleware, middleware)
    },
  }
}
