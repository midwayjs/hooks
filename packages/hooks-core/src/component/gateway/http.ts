import { existsSync } from 'fs'
import staticCache from 'koa-static-cache'
import {
  HTTPRoute,
  ServerRoute,
} from 'packages/hooks-core/dist/midwayjs-hooks-core.cjs'
import { __decorate } from 'tslib'
import { join } from 'upath'

import {
  IMidwayApplication,
  IMidwayContainer,
  IMidwayContext,
} from '@midwayjs/core'
import { All, Controller, Inject, Provide } from '@midwayjs/decorator'

import { superjson } from '../../lib'
import { useContext } from '../../runtime'
import { ApiFunction } from '../../types/common'
import {
  ComponentConfig,
  CreateApiParam,
  HooksGatewayAdapter,
} from '../../types/gateway'
import { isDevelopment, useHooksMiddleware } from '../../util'

export class HTTPGateway implements HooksGatewayAdapter {
  config: ComponentConfig
  container: IMidwayContainer
  app: IMidwayApplication<IMidwayContext>

  constructor(config: ComponentConfig) {
    this.config = config
  }

  is(route: ServerRoute<HTTPRoute>) {
    return !!route?.basePath
  }

  createApi(param: CreateApiParam) {
    const { id, fn, httpPath } = param

    // 设置中间件
    fn.middleware.unshift(...this.getMiddleware())

    // Source: https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&importHelpers=true&emitDecoratorMetadata=false&ts=4.1.5#code/JYWwDg9gTgLgBAbzgSQHYCsCmBjGAaOAYQlRiggBsLMoCBBKggBXIDdgATTOAXzgDNyIOAHIAAiE4B3AIYBPdAGcA9F2zQZMaCIBQO9akXx+qOAF44ACgB0tmVADmigFxwZqOQG0AugEpzAHyIPHpiLBDsXJa+OmLEpORUNJYiyiIx2BQyiopwAGIArqi4wCTxMDLAqDSIOnD1cGJoWLjRdQ24AB6u7nJ6DY0MFCkiBEiSHBzUslCYrj68MQPZcsVwABbuUzXRtQP11PD2TuZwMOvAitZd1rMAjgWYRtYARhAcctbHuQA+P3A+dr7OCzGAFKCmGSyYDGVA2OyORRLBohEJAA
    let FunctionContainer = class FunctionContainer {
      ctx: any
      async handler() {
        let args = this.ctx.request?.body?.args || []
        return await fn(...args)
      }
    }
    __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
    __decorate(
      [All(httpPath, { middleware: fn.middleware })],
      FunctionContainer.prototype,
      'handler',
      null
    )
    FunctionContainer = __decorate(
      [Provide(id), Controller('/')],
      FunctionContainer
    )

    this.container.bind(id, FunctionContainer)
  }

  getMiddleware() {
    const mws = [this.handleError]

    const enableSuperjson = this.config.internal.superjson
    if (enableSuperjson) {
      mws.push(this.deserializeSuperjson)
    }

    return mws.map(useHooksMiddleware)
  }

  async deserializeSuperjson(next: any) {
    await next()
    const ctx = useContext()
    if (ctx.type.includes('application/json')) {
      ctx.body = superjson.serialize(ctx.body)
    }
  }

  handleError = async (next: any) => {
    try {
      await next()
    } catch (error) {
      const ctx = useContext()

      if (this.config.internal.superjson) {
        ctx.status = 500
        ctx.body = superjson.serialize(error)
      } else {
        throw error
      }
    }
  }

  afterCreate() {
    if (isDevelopment()) {
      return
    }

    if (!this.isViteProject) {
      return
    }

    const {
      router: { routes },
    } = this.config
    if (routes.has('/') || routes.has('/*')) {
      return
    }

    const baseDir = this.container.get('baseDir')
    const mw = staticCache({
      dir: join(baseDir, '..', this.config.internal.build.viteOutDir),
      dynamic: true,
      alias: {
        '/': 'index.html',
        /**
         * Add alias for windows, '/' -> '\\'
         * https://github.com/koajs/static-cache/blob/master/index.js#L45
         */
        '\\': 'index.html',
      },
      buffer: true,
      gzip: true,
    })

    const fn: ApiFunction = async () => {}
    fn.middleware = [mw]

    this.createApi({
      id: 'hooks-host',
      httpPath: '/*',
      fn,
    })
  }

  get isViteProject() {
    return ['vite.config.ts', 'vite.config.js'].some((config) =>
      existsSync(join(this.config.root, config))
    )
  }
}
