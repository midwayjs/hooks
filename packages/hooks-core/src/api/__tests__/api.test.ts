import { Api } from '..'
import { extractMetadata } from '../../common/util'
import { Get, Headers, Params, Post, Query } from '../operator/http'
import { Middleware } from '../operator/middleware'
import { Operator } from '../type'

it('Api should work', async () => {
  const hello = Api(async () => {
    return 'Hello Api!'
  })
  expect(await hello()).toEqual('Hello Api!')

  const withArgs = Api(async (name: string) => {
    return name
  })
  expect(await withArgs('hooks')).toEqual('hooks')

  const withOperator = Api(Post(), async (name: string) => {
    return name
  })
  expect(await withOperator('hooks')).toEqual('hooks')
})

it('compose with meta operator', async () => {
  const withQuery = Api(Get(), Query<{ name: string }>(), async () => {
    return 'Hello World'
  })
  expect(await withQuery({ query: { name: 'hooks' } })).toEqual('Hello World')
})

it('defineMeta', async () => {
  const logger = () => {}
  const fn = Api(
    Get(),
    Query(),
    Params(),
    Headers(),
    Middleware(logger),
    async () => {}
  )

  expect(extractMetadata(fn)).toMatchSnapshot()
})

it('execute', async () => {
  const stack = []
  const CustomExecutor = (): Operator<void> => {
    return {
      name: 'CustomExecutor',
      async execute(helper, next) {
        stack.push(1)
        await next()
        stack.push(4)
      },
    }
  }

  const fn = Api(CustomExecutor(), async (...numbers: number[]) => {
    stack.push(...numbers)
    return numbers
  })

  const result = await fn(2, 3)
  expect(stack).toEqual([1, 2, 3, 4])
  expect(result).toEqual([2, 3])
})
