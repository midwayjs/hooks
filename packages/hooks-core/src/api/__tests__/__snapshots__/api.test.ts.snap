// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineMeta 1`] = `
Object {
  "Http_Header": true,
  "Http_Params": true,
  "Http_Query": true,
  "Middleware": Array [
    [Function],
  ],
  "Trigger": Object {
    "method": "GET",
    "path": undefined,
    "requestClient": Object {
      "client": "@midwayjs/rpc",
      "fetcher": "http",
    },
    "type": "HTTP",
  },
  "USE_INPUT_METADATA": true,
}
`;
